File: lib/AnalogPulse.cpp
/**
 * @file lib/AnalogPulse.cpp
 * @author Guilherme Inacio Goncalves <ggoncalves@iprj.uer.br>
 * @copyright Copyright 2022, Guilherme Inacio Goncalves. All rights reserved.
 *
 * This file is part of CPS (Calorimetry Pulse Simulator).
 *
 * CPS is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or any later version.
 *
 * CPS is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * CPS. If not, see <https://www.gnu.org/licenses/>.
 */
#include "AnalogPulse.h"
#include <cmath>

using namespace cps;

AnalogPulse::AnalogPulse(
    const IPulseShape* pulseShape,
    double amplitude,
    double pedestal,
    double phase,
    double deformationLevel,
    double noiseMean,
    double noiseStdDev
    ):
  m_pulseShape(pulseShape),
  m_amplitude(amplitude),
  m_pedestal(pedestal),
  m_phase(phase),
  m_deformationLevel(deformationLevel),
  m_noiseMean(noiseMean),
  m_noiseStdDev(noiseStdDev),
  m_generator()
{}

AnalogPulse::AnalogPulse(const AnalogPulse& source):
  m_pulseShape(source.m_pulseShape),
  m_amplitude(source.m_amplitude),
  m_phase(source.m_phase),
  m_pedestal(source.m_pedestal),
  m_deformationLevel(source.m_deformationLevel),
  m_noiseMean(source.m_noiseMean),
  m_noiseStdDev(source.m_noiseStdDev),
  m_generator()
{}

bool AnalogPulse::operator==(const AnalogPulse& rhs) const {
  return m_pulseShape == rhs.m_pulseShape
    && m_amplitude == rhs.m_amplitude
    && m_pedestal == rhs.m_pedestal
    && m_phase == rhs.m_phase
    && m_deformationLevel == rhs.m_deformationLevel
    && m_noiseMean == rhs.m_noiseMean
    && m_noiseStdDev == rhs.m_noiseStdDev;
}

bool AnalogPulse::operator!=(const AnalogPulse& rhs) const {
  return !(*this == rhs);
}

double AnalogPulse::GetSample(const double& time) const {
  double timeWithPhaseDeviation = time + m_phase;
  double shapeSample            = m_pulseShape->GetY(timeWithPhaseDeviation);
  double deformation            = GenerateDeformation(shapeSample);
  double noise                  = GenerateNoise();
  return m_amplitude * (shapeSample + deformation) + m_pedestal + noise;
}

double AnalogPulse::GenerateDeformation(double shapeSample) const {
  if (m_deformationLevel == 0) return 0;
  double stdDev = m_deformationLevel * std::abs(shapeSample);
  std::normal_distribution<double> distribution(0, stdDev);
  return distribution(m_generator);
}

double AnalogPulse::GenerateNoise() const {
  if (m_noiseMean == 0 && m_noiseStdDev == 0) return 0;
  std::normal_distribution<double> distribution(m_noiseMean, m_noiseStdDev);
  return distribution(m_generator);
}

File: docs/examples/python/analog_pulse.py
#!/usr/bin/env python3

# @file docs/examples/python/analog_pulse.py
# @author Guilherme Inacio Goncalves <ggoncalves@iprj.uer.br>
# @copyright Copyright 2022, Guilherme Inacio Goncalves. All rights reserved.
#
# This file is part of CPS (Calorimetry Pulse Simulator).
#
# CPS is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or any later version.
#
# CPS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# CPS. If not, see <https://www.gnu.org/licenses/>.

import matplotlib.pyplot as plt
import numpy as np
from pycps import AnalogPulse, TextFilePulseShape, Digitizer


def main():
    """
    main function
    """

    shape_path = "../../../tests/fixtures/unipolar-pulse-shape.dat"
    pulse_shape = TextFilePulseShape(shape_path)
    digitizer = Digitizer(7, 25, -75)

    amplitude = 100.0
    pedestal = 0.0

    default_pulse = AnalogPulse(pulse_shape, amplitude)
    plot_pulse(default_pulse, digitizer, "default pulse")

    phase = -25
    phased_pulse = AnalogPulse(pulse_shape, amplitude, pedestal, phase)
    plot_pulse(phased_pulse, digitizer, "phased pulse")

    phase = 25
    noise_mean = 0
    noise_sigma = 1.5
    noisy_and_phased_pulse = AnalogPulse(
        pulse_shape,
        amplitude,
        pedestal,
        phase,
        noise_mean,
        noise_sigma
    )
    plot_pulse(noisy_and_phased_pulse, digitizer, "noisy and phased pulse")

    plt.grid(zorder=0, linestyle='--')
    plt.legend()
    plt.xlabel("time", horizontalalignment='right', x=1.0, zorder=3)
    plt.ylabel("ADC", horizontalalignment='right', y=1.0)
    plt.show()


def plot_pulse(pulse, digitizer, label):
    """
    draw a pulse
    """
    time_series = list()
    sample_series = list()
    for time in np.arange(-75.5, 124.5, 0.5):
        time_series.append(time)
        sample_series.append(pulse.get_sample(time))
    plot = plt.plot(time_series, sample_series, "-", label=label)
    digital_samples_time = [-75.0, -50.0, -25.0, 0.0, 25.0, 50.0, 75.0]
    digitial_samples = digitizer.digitize(pulse)
    plt.plot(digital_samples_time, digitial_samples, ".",
             color=plot[0].get_color(),
             markersize=10,
             markeredgewidth=1.5,
             markeredgecolor="black")


if __name__ == '__main__':
    main()
